modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -485,58 +283,57 @@ private SingleRowResponse handleGetAndUpsertCommand(GetAndUpsertCommand cmd, lon
-
-            LockManager lockManager = txManager.lockManager();
-
-            // This code is technically incorrect and assumes that "stateChanged" is always true.
-            // This was done because transaction state is not persisted
-            // and thus FinishTxCommand couldn't be completed on recovery after node restart ("changeState" uses "replace").
-            if (/*txManager.state(txId) == TxState.COMMITED*/cmd.finish()) {
-                lockManager.locks(txId)
-                        .forEachRemaining(
-                                lock -> {
-                                    storage.commitWrite((ByteBuffer) lock.lockKey().key(), txId);
-                                }
-                        );
-            } else /*if (txManager.state(txId) == TxState.ABORTED)*/ {
-                lockManager.locks(txId)
-                        .forEachRemaining(
-                                lock -> {
-                                    storage.abortWrite((ByteBuffer) lock.lockKey().key());
-                                }
-                        );
-            }
+        UUID txId = cmd.getTxId();
+
+        boolean stateChanged = txManager.changeState(txId, TxState.PENDING, cmd.finish() ? TxState.COMMITED : TxState.ABORTED);
+
+        LockManager lockManager = txManager.lockManager();
+
+        // This code is technically incorrect and assumes that "stateChanged" is always true.
+        // This was done because transaction state is not persisted
+        // and thus FinishTxCommand couldn't be completed on recovery after node restart ("changeState" uses "replace").
+        if (/*txManager.state(txId) == TxState.COMMITED*/cmd.finish()) {
+            lockManager.locks(txId)
+                    .forEachRemaining(
+                            lock -> {
+                                storage.commitWrite((RowId) lock.lockKey().key(), new Timestamp(txId));
+                            }
+                    );
+        } else /*if (txManager.state(txId) == TxState.ABORTED)*/ {
+            lockManager.locks(txId)
+                    .forEachRemaining(
+                            lock -> {
+                                storage.abortWrite((RowId) lock.lockKey().key());
+                            }
+                    );
+        }
 
-            // TODO: tmp
-            if (/*txManager.state(txId) == TxState.COMMITED*/cmd.finish()) {
-                storage.pendingKeys.getOrDefault(txId, Collections.emptyList()).forEach(key -> storage.commitWrite((ByteBuffer) key, txId));
-            } else /*if (txManager.state(txId) == TxState.ABORTED)*/ {
-                storage.pendingKeys.getOrDefault(txId, Collections.emptyList()).forEach(key -> storage.abortWrite((ByteBuffer) key));
-            }
+        // TODO: tmp
+        if (/*txManager.state(txId) == TxState.COMMITED*/cmd.finish()) {
+            //storage.pendingKeys.getOrDefault(txId, Collections.emptyList())
```

Comment:
Why is this commented?

Reply:
It will work after transactional commit is implemented.